krm-functions/lib/nad/v1/nad.go

```diff
@@ -0,0 +1,366 @@
+func NewFromKubeObject(b *fn.KubeObject) (*NadStruct, error) {
+	p, err := kubeobject.NewFromKubeObject[nadv1.NetworkAttachmentDefinition](b)
+	if err != nil {
+		return nil, err
+	}
+	return &NadStruct{K: *p}, nil
+}
+
+// NewFromYAML creates a new parser interface
+// It expects a raw byte slice as input representing the serialized yaml file
+func NewFromYAML(b []byte) (*NadStruct, error) {
+	p, err := kubeobject.NewFromYaml[nadv1.NetworkAttachmentDefinition](b)
+	if err != nil {
+		return nil, err
+	}
+	return &NadStruct{K: *p}, nil
+}
+
+// NewFromGoStruct creates a new parser interface
+// It expects a go struct representing the interface krm resource
+func NewFromGoStruct(b *nadv1.NetworkAttachmentDefinition) (*NadStruct, error) {
+	p, err := kubeobject.NewFromGoStruct(b)
+	if err != nil {
+		return nil, err
+	}
+	return &NadStruct{K: *p}, nil
+}
+
+func (r *NadStruct) GetStringValue(fields ...string) string {
+	s, ok, err := r.K.NestedString(fields...)
+	if err != nil {
+		return ""
+	}
+	if !ok {
+		return ""
+	}
+	return s
+}
+
+func (r *NadStruct) GetConfigSpec() string {
+	return r.GetStringValue(ConfigType...)
+}
+
+func (r *NadStruct) GetCNIType() (string, error) {
+	existingNadConfig, err := r.GetNadConfig()
+	if err != nil {
+		return "", err
+	}
+	for _, plugin := range existingNadConfig.Plugins {
+		if plugin.Type == TuningType {
+			continue
+		} else {
+			return plugin.Type, nil
+		}
+	}
+	return "", nil
+}
+
+func (r *NadStruct) GetVlan() (int, error) {
+	existingNadConfig, err := r.GetNadConfig()
+	if err != nil {
+		return 0, err
+	}
+	return existingNadConfig.Vlan, nil
+}
+
+func (r *NadStruct) GetNadMaster() (string, error) {
+	existingNadConfig, err := r.GetNadConfig()
+	if err != nil {
+		return "", err
+	}
+	for _, plugin := range existingNadConfig.Plugins {
+		if plugin.Type == TuningType {
+			continue
+		} else {
+			return plugin.Master, nil
+		}
+	}
+	return "", nil
+}
+
+func (r *NadStruct) GetIpamAddress() ([]Addresses, error) {
+	existingNadConfig, err := r.GetNadConfig()
+	if err != nil {
+		return []Addresses{}, err
+	}
+	for _, plugin := range existingNadConfig.Plugins {
+		if plugin.Type == TuningType {
+			continue
+		} else {
+			return plugin.Ipam.Addresses, nil
+		}
+	}
+	return []Addresses{}, nil
+}
+
+func (r *NadStruct) GetNadConfig() (NadConfig, error) {
+	nadConfigStruct := NadConfig{}
+	configSpec := r.GetConfigSpec()
+	if configSpec == "" {
+		configSpec = "{}"
+	}
+	if err := json.Unmarshal([]byte(configSpec), &nadConfigStruct); err != nil {
+		return nadConfigStruct, fmt.Errorf("invalid NAD Config, %s", err)
+	}
+	nadConfigStruct.CniVersion = CniVersion
+	if r.CniSpecType == VlanAllocOnly {
+		return nadConfigStruct, nil
+	} else if r.CniSpecType == IpVlanType {
+		if nadConfigStruct.Plugins == nil || len(nadConfigStruct.Plugins) == 0 {
+			nadConfigStruct.Plugins = []PluginCniType{
+				{
+					Capabilities: Capabilities{
```

Comment:
Not sure on the convention here but there seems to be some inconsistencies in the formatting/indentation.

Capabilities: Capabilities{
						Ips: true,
					}

vs

Capabilities: Capabilities{Ips: true}